### PR TITLE
Add events for email otp related scenarios

### PR DESCRIPTION
**Pull request by lakshani**
Add events for email otp generate and validate flows.

Fix for https://github.com/wso2/product-is/issues/10706

### CODE REVIEW

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 1]
```diff
@@ -2519,4 +2528,133 @@ private long getUnlockTimeInMilliSeconds(AuthenticatedUser authenticatedUser) th
...
+                .AUTHENTICATED_USER);
+        eventProperties.put(IdentityEventConstants.EventProperty.USER_NAME, authenticatedUser.getUserName());
+        eventProperties.put(IdentityEventConstants.EventProperty.TENANT_DOMAIN, context.getTenantDomain());
+        eventProperties.put(IdentityEventConstants.EventProperty.USER_STORE_DOMAIN, authenticatedUser
+                .getUserStoreDomain());
+        eventProperties.put(IdentityEventConstants.EventProperty.APPLICATION_NAME, context.getServiceProviderName());
+        eventProperties.put(IdentityEventConstants.EventProperty.USER_AGENT, request.getHeader(
+                EmailOTPAuthenticatorConstants.USER_AGENT));
+
+        if (StringUtils.isNotEmpty(request.getHeader(EmailOTPAuthenticatorConstants.X_FORWARDED_FOR))) {
```
Comment (madurangasiriwardena) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: Use `getClientIpAddress` from `IdentityUtil`

Comment (lakshani) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: Fixed in 7ad7f37f8bbf44021015f9a6b207cdf0219f076a

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 2]
```diff
@@ -2519,4 +2528,115 @@ private long getUnlockTimeInMilliSeconds(AuthenticatedUser authenticatedUser) th
...
+            if (log.isDebugEnabled()) {
+                log.debug("Setting false resend-code property in event since http request has not resendCode parameter.");
+            }
+            eventProperties.put(IdentityEventConstants.EventProperty.RESEND_CODE, false);
+        }
+
+        eventProperties.put(IdentityEventConstants.EventProperty.GENERATED_OTP, context.getProperty(
+                EmailOTPAuthenticatorConstants.OTP_TOKEN));
+
+
```
Comment (pamodaaw) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: unnecessary new line

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 3]
```diff
@@ -2519,4 +2528,115 @@ private long getUnlockTimeInMilliSeconds(AuthenticatedUser authenticatedUser) th
...
+        AuthenticatedUser authenticatedUser = (AuthenticatedUser) context.getProperty(EmailOTPAuthenticatorConstants
+                .AUTHENTICATED_USER);
+        eventProperties.put(IdentityEventConstants.EventProperty.USER_NAME, authenticatedUser.getUserName());
+        eventProperties.put(IdentityEventConstants.EventProperty.TENANT_DOMAIN, context.getTenantDomain());
+        eventProperties.put(IdentityEventConstants.EventProperty.USER_STORE_DOMAIN, authenticatedUser
+                .getUserStoreDomain());
+        eventProperties.put(IdentityEventConstants.EventProperty.APPLICATION_NAME, context.getServiceProviderName());
+        eventProperties.put(IdentityEventConstants.EventProperty.USER_AGENT, request.getHeader(
+                EmailOTPAuthenticatorConstants.USER_AGENT));
+        if (request.getParameter(EmailOTPAuthenticatorConstants.RESEND) != null) {
```
Comment (pamodaaw) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: Shall we do a StringUtils.isNotBlank function here

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 4]
```diff
@@ -2519,4 +2528,115 @@ private long getUnlockTimeInMilliSeconds(AuthenticatedUser authenticatedUser) th
...
+        long otpGeneratedTime = (long) context.getProperty(EmailOTPAuthenticatorConstants.OTP_GENERATED_TIME);
+        eventProperties.put(IdentityEventConstants.EventProperty.OTP_GENERATED_TIME, otpGeneratedTime);
+
+        long expiryTime = otpGeneratedTime + Long.parseLong(getExpireTime(context));
+        eventProperties.put(IdentityEventConstants.EventProperty.OTP_EXPIRY_TIME, expiryTime);
+        Object codeMismatch = context.getProperty(EmailOTPAuthenticatorConstants.CODE_MISMATCH);
+        Object otpExpired = context.getProperty(EmailOTPAuthenticatorConstants.OTP_EXPIRED);
+
+        String status;
+        if (otpExpired != null && (String) otpExpired == EmailOTPAuthenticatorConstants.TRUE) {
```
Comment (pamodaaw) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java:2621]: Use boolean functions

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 5]
```diff
@@ -2519,4 +2528,115 @@ private long getUnlockTimeInMilliSeconds(AuthenticatedUser authenticatedUser) th
...
+            status = EmailOTPAuthenticatorConstants.STATUS_OTP_EXPIRED;
+        } else if (codeMismatch != null && (Boolean) codeMismatch) {
```
Comment (pamodaaw) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java:2623]: use boolean functions

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticatorConstants.java [hunk 6]
```diff
@@ -127,6 +127,13 @@
     public static final String SERVICE_PROVIDER_NAME = "serviceProviderName";
     public static final String IP_ADDRESS = "ipAddress";
 
+   public static final String USER_AGENT = "user-agent";
```
Comment (pamodaaw) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticatorConstants.java]: formatting issue

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java (hunk 5 above)
Comment (lakshani) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java:2623]: Boolean function only supports for Strings ,not for objects

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java (hunk 4 above)
Comment (lakshani) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java:2621]: Boolean function only supports for Strings, not for objects

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java (hunk 2 above)
Comment (lakshani) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: Fixed in 5a5d198b3054bf710058f1c945a1f73b350bd783

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java (hunk 3 above)
Comment (lakshani) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: Fixed in 5a5d198b3054bf710058f1c945a1f73b350bd783

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticatorConstants.java (hunk 6 above)
Comment (lakshanigamage) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticatorConstants.java]: Fixed in https://github.com/wso2-extensions/identity-outbound-auth-email-otp/pull/94/commits/5a5d198b3054bf710058f1c945a1f73b350bd783

Comment (lakshani) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticatorConstants.java]: Fixed in https://github.com/wso2-extensions/identity-outbound-auth-email-otp/pull/94/commits/5a5d198b3054bf710058f1c945a1f73b350bd783

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 7]
```diff
@@ -2519,4 +2528,113 @@ private long getUnlockTimeInMilliSeconds(AuthenticatedUser authenticatedUser) th
...
```
Comment (ashensw) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java:2623]:
```suggestion
        boolean codeMismatch = Boolean.valueOf(context.getProperty(EmailOTPAuthenticatorConstants.CODE_MISMATCH));
        boolean otpExpired = Boolean.valueOf(context.getProperty(EmailOTPAuthenticatorConstants.OTP_EXPIRED));

        String status;
        if (otpExpired) {
            status = EmailOTPAuthenticatorConstants.STATUS_OTP_EXPIRED;
        } else if (codeMismatch) {
```

Comment (ashensw) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java:2623]: Shall we refactor the code as above? WDYT?

**Review by ashensw — APPROVED** [review]

**Pull request merged**